### PR TITLE
New style negative_sampling

### PR DESCRIPTION
**Pull request by hvy**
New style negative_sampling to support double backprop.

### CODE REVIEW

**Comment by hvy** [comment]
Rebased on top of current master to resolve conflicts.

File: chainer/functions/loss/negative_sampling.py [hunk 1]
```diff
@@ -104,20 +114,44 @@ def forward_gpu(self, inputs):
...
+        grad = NegativeSamplingFunctionGrad(
+            self.reduce, self.ignore_mask, self.sample_size, self.samples,
+            self.wx)
+        grad.ignore_label = self.ignore_label
+        return grad.apply((x, W, gy))
+
+
+class NegativeSamplingFunctionGrad(function_node.FunctionNode):
+
+    ignore_label = -1
```
Comment (takagi) [inline chainer/functions/loss/negative_sampling.py]: `ignore_label` looks not used in the grad function.

Comment (hvy) [reply chainer/functions/loss/negative_sampling.py]: Thanks, I removed the variable and the assignment in the main class.

**Comment by hvy** [comment]
Thank you for the review. I changed to code accordingly.

**Comment by hvy** [comment]
Should this PR be backported?

**Comment by takagi** [comment]
I made the backport of this PR as other new style functions.

**Comment by hvy** [comment]
Ah #4021 , thanks!

**Pull request merged**